### PR TITLE
Let gradient checkpointing skip layers with every_n_layers

### LINKED ISSUE #26103: Add gradient_checkpointing_segment_size
### Feature request

Currently when we enable gradient checkpointing, e.g. in `LlamaModel`, we call `torch.utils.checkpoint.checkpoint` on every `LlamaDecoderLayer`. As per [Training Deep Nets with Sublinear Memory Cost](https://arxiv.org/pdf/1604.06174.pdf), https://github.com/cybertronai/gradient-checkpointing and hinted by [torch.utils.checkpoint.checkpoint_sequential](https://pytorch.org/docs/stable/checkpoint.html#torch.utils.checkpoint.checkpoint_sequential), we can strike a balance between memory and compute by supporting `gradient_checkpointing_segment_size`.

### Motivation

Gradient checkpointing makes training slow and people enable it mainly to avoid VRAM OOM. It would be great if we can leverage `gradient_checkpointing_segment_size` and make gradient checkpointing configurable.

### Your contribution

Here is a diff to start off with:

```diff
diff --git a/src/transformers/modeling_utils.py b/src/transformers/modeling_utils.py
index bbf229b58..685400ffd 100644
--- a/src/transformers/modeling_utils.py
+++ b/src/transformers/modeling_utils.py
@@ -1745,6 +1745,16 @@ class PreTrainedModel(nn.Module, ModuleUtilsMixin, GenerationMixin, PushToHubMix
         """
         return any(hasattr(m, "gradient_checkpointing") and m.gradient_checkpointing for m in self.modules())
 
+    def set_gradient_checkpointing_segment_size(self, value: int):
+        """
+        Set gradient checkpointing segment size for the current model.
+        """
+        if not self.supports_gradient_checkpointing:
+            raise ValueError(f"{self.__class__.__name__} does not support gradient checkpointing.")
+        if not hasattr(self.model, "_set_gradient_checkpointing_segment_size"):
+            raise ValueError(f"{self.__class__.__name__} does not support configuring gradient checkpointing segment size.")
+        self.apply(partial(self._set_gradient_checkpointing_segment_size, value=value))
+
     def save_pretrained(
         self,
         save_directory: Union[str, os.PathLike],
diff --git a/src/transformers/models/llama/modeling_llama.py b/src/transformers/models/llama/modeling_llama.py
index 5e7a879c0..682e49b21 100644
--- a/src/transformers/models/llama/modeling_llama.py
+++ b/src/transformers/models/llama/modeling_llama.py
@@ -491,6 +491,10 @@ class LlamaPreTrainedModel(PreTrainedModel):
         if isinstance(module, LlamaModel):
             module.gradient_checkpointing = value
 
+    def _set_gradient_checkpointing_segment_size(self, module, value=1):
+        if isinstance(module, LlamaModel):
+            module.gradient_checkpointing_segment_size = value
+
 
 LLAMA_INPUTS_DOCSTRING = r"""
     Args:
@@ -578,6 +582,7 @@ class LlamaModel(LlamaPreTrainedModel):
         self.norm = LlamaRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
 
         self.gradient_checkpointing = False
+        self.gradient_checkpointing_segment_size = 1
         # Initialize weights and apply final processing
         self.post_init()
 
@@ -689,7 +694,7 @@ class LlamaModel(LlamaPreTrainedModel):
 
             past_key_value = past_key_values[idx] if past_key_values is not None else None
 
-            if self.gradient_checkpointing and self.training:
+            if self.gradient_checkpointing and self.training and (idx % self.gradient_checkpointing_segment_size == 0):
 
                 def create_custom_forward(module):
                     def custom_forward(*inputs):
diff --git a/src/transformers/trainer.py b/src/transformers/trainer.py
index 2b8cb013f..f7d5e87ea 100755
--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -1687,6 +1687,9 @@ class Trainer:
         if args.gradient_checkpointing:
             self.model.gradient_checkpointing_enable()
 
+            if args.gradient_checkpointing_segment_size > 1:
+               self.model.set_gradient_checkpointing_segment_size(args.gradient_checkpointing_segment_size)
+
         model = self._wrap_model(self.model_wrapped)
 
         if (is_sagemaker_mp_enabled() or self.is_fsdp_enabled) and resume_from_checkpoint is not None:
diff --git a/src/transformers/training_args.py b/src/transformers/training_args.py
index 2f72622d9..ecffc5660 100644
--- a/src/transformers/training_args.py
+++ b/src/transformers/training_args.py
@@ -586,6 +586,8 @@ class TrainingArguments:
             Unless this is `True`, the `Trainer` will skip pushing a checkpoint when the previous push is not finished.
         gradient_checkpointing (`bool`, *optional*, defaults to `False`):
             If True, use gradient checkpointing to save memory at the expense of slower backward pass.
+        gradient_checkpointing_segment_size (`int`, *optional*, defaults to `1`):
+            If gradient_checkpointing is True, use gradient checkpointing for every segment size
         include_inputs_for_metrics (`bool`, *optional*, defaults to `False`):
             Whether or not the inputs will be passed to the `compute_metrics` function. This is intended for metrics
             that need inputs, predictions and references for scoring calculation in Metric class.
@@ -1144,6 +1146,12 @@ class TrainingArguments:
             "help": "If True, use gradient checkpointing to save memory at the expense of slower backward pass."
         },
     )
+    gradient_checkpointing_segment_size: int = field(
+        default=1,
+        metadata={
+            "help": "If gradient_checkpointing is True, use gradient checkpointing for every segment size."
+        },
+    )
     include_inputs_for_metrics: bool = field(
         default=False, metadata={"help": "Whether or not the inputs will be passed to the `compute_metrics` function."}
     )
@@ -2137,6 +2145,7 @@ class TrainingArguments:
         gradient_accumulation_steps: int = 1,
         seed: int = 42,
         gradient_checkpointing: bool = False,
+        gradient_checkpointing_segment_size: int = 1,
     ):
         """
         A method that regroups all basic arguments linked to the training.
@@ -2179,6 +2188,8 @@ class TrainingArguments:
                 parameters.
             gradient_checkpointing (`bool`, *optional*, defaults to `False`):
                 If True, use gradient checkpointing to save memory at the expense of slower backward pass.
+            gradient_checkpointing_segment_size (`int`, *optional*, defaults to `1`):
+                If gradient_checkpointing is True, use gradient checkpointing for every segment size
 
         Example:
 
```

### PR DESCRIPTION
**Pull request by qgallouedec**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48200) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48200)

Closes #26103.

Gradient checkpointing is all-or-nothing today: `gradient_checkpointing=True` recomputes every decoder layer.
That is right when memory is the binding constraint and wasteful when it is not, and once the loss and the kernels stop dominating memory, it often is not.

This adds an `every_n_layers` parameter to `gradient_checkpointing_enable()`, so the memory that full checkpointing frees can be spent back on speed:

```python
model.gradient_checkpointing_enable(every_n_layers=4)

# or through the Trainer, with no new argument:
TrainingArguments(gradient_checkpointing=True,
                  gradient_checkpointing_kwargs={"every_n_layers": 4})
```

`1` is the current behaviour and stays the default, so nothing changes for anyone who does not ask for it.
Only `GradientCheckpointingLayer` instances are counted, so the ratio means what it says even when other modules carry a `gradient_checkpointing` flag.

## Why now? the 2023 blocker is gone

This was proposed in #26103 back in September 2023 and it was deferred, and the reasons were:

> we try to keep our forward passes as simple as possible. As gradient checkpointing is used by most of our models, **it is a change which would have to be applied across the library**, and so requires careful consideration of how it's configured e.g. **would this lead to many additional future arguments?** And backwards compatibility.

Both objections are answerable today:

- **"applied across the library"** was true when each model hand-wrote its own layer loop. Since `GradientCheckpointingLayer` was introduced, that dispatch is centralised; this PR touches `modeling_utils.py`, `trainer.py` and a docstring, and **no model file at all**.
- **"many additional future arguments"**: this adds none. `every_n_layers` is read out of the existing `gradient_checkpointing_kwargs`, and every other key in that dict is still forwarded to `torch.utils.checkpoint.checkpoint` as before.
- **backwards compatibility**, the default is `1`, which is today's behaviour.

Also @grimulkan noted that the original 2023 diff would run the skipped layers through a plain `forward()`, "which will
not be with `no_grad()` or using the correct non-reentrant logic". That does not apply here: this PR only sets the existing `gradient_checkpointing` flag per layer, and `GradientCheckpointingLayer.__call__` already handles both paths. A skipped layer takes exactly the code path it takes today with `gradient_checkpointing=False`.

## Numbers

`Qwen3-0.6B`, seq 4096, per-device batch 2, bf16, 1×H100:

fig_gc

| | throughput | peak memory |
|---|---|---|
| all 28 layers checkpointed (today) | 27605 tokens/s | 17.81 GB |
| `every_n_layers=2` | **30429** (1.10×) | 25.26 GB |
| `every_n_layers=4` | **32061** (1.16×) | 28.99 GB |
| `every_n_layers=8` | **32809** (1.19×) | 30.59 GB |
| `gradient_checkpointing=False` | 34080 (1.23×) | 32.72 GB |

The effect is larger on models where recompute is a bigger share of the step; on a 40-layer MoE with 30 gated-delta-net layers I measured 1.45× at `n=8`.

<details>
<summary>bench_gc.py — the script that produced the table</summary>

```python
# bench_gc.py — step time and peak memory vs how many layers are checkpointed.
import statistics, sys, time

import numpy as np
import torch
from datasets import Dataset
from transformers import AutoConfig, AutoModelForCausalLM, Trainer, TrainerCallback, TrainingArguments

MODEL, SEQ, BS, STEPS, WARMUP = "Qwen/Qwen3-0.6B", 4096, 2, 12, 4


class Timer(TrainerCallback):
    def __init__(self):
        self.times, self.t = [], None

    def on_step_begin(self, *a, **k):
        torch.cuda.synchronize()
        self.t = time.perf_counter()

    def on_step_end(self, args, state, control, **k):
        torch.cuda.synchronize()
        if state.global_step > WARMUP:
            self.times.append(time.perf_counter() - self.t)


def collate(rows):
    ids = torch.tensor([r["input_ids"] for r in rows])
    return {"input_ids": ids, "attention_mask": torch.ones_like(ids), "labels": ids}


vocab = AutoConfig.from_pretrained(MODEL).vocab_size
rng = np.random.default_rng(0)
ds = Dataset.from_dict({"input_ids": rng.integers(0, vocab, size=(BS * (STEPS + 2), SEQ)).tolist()})

for n in (1, 2, 4, 8, None):  # None = gradient_checkpointing=False, the ceiling
    model = AutoModelForCausalLM.from_pretrained(MODEL, dtype=torch.bfloat16).cuda()
    timer = Timer()
    args = TrainingArguments(
        output_dir="/tmp/gcbench", per_device_train_batch_size=BS, max_steps=STEPS,
        gradient_checkpointing=n is not None,
        gradient_checkpointing_kwargs=None if n is None else {"every_n_layers": n},
        bf16=True, logging_steps=1, report_to=[], save_strategy="no", seed=0,
    )
    torch.cuda.reset_peak_memory_stats()
    Trainer(model=model, args=args, train_dataset=ds, data_collator=collate, callbacks=[timer]).train()
    step = statistics.median(timer.times)
    label = "checkpointing off" if n is None else f"every_n_layers={n}"
    print(f"{label:>18}: {BS * SEQ / step:8.0f} tokens/s   {torch.cuda.max_memory_allocated() / 2**30:5.2f} GB peak")
    del model
    torch.cuda.empty_cache()
```

</details>

### CODE REVIEW

**Review by SunMarc — APPROVED** [review]
Thanks for the clean integration !
